File: 프로그래머스/4/42897. 도둑질/도둑질.cpp
#include <string>
#include <vector>
#include <algorithm>

using namespace std;

int dp[1000006];
int solution(vector<int> money) {
    int ans = 0, n = money.size();
    // 0 방문 n - 1 x
    dp[0] = money[0];
    dp[1] = max(money[0], money[1]);
    
    for(int i = 2; i < n - 1; i++)
        dp[i] = max(dp[i-1], dp[i-2] + money[i]);
    
    ans = dp[n-2];
    
    // n - 1 방문 0 x
    fill(dp, dp + n, 0);
    dp[1] = money[1];
    
    for(int i = 2; i < n; i++)
        dp[i] = max(dp[i-1], dp[i-2] + money[i]);
   
    ans = max(ans, dp[n-1]);
    
    return ans;
}

File: 프로그래머스/4/42897. 도둑질/도둑질.java
import java.util.*;

class Solution {
    public int solution(int[] money) {
        int n = money.length;
        int[] first = new int[n];
        int[] second = new int[n];
        
        for(int i = 0; i < n; i++) {
            first[i] = money[i];
            second[i] = money[i];
        }
        
        first[1] = second[0] = -1;
        first[2] += first[0];
        for(int i = 3; i < n; i++){
            first[i] += Math.max(first[i - 2], first[i - 3]);
            second[i] += Math.max(second[i - 2], second[i - 3]);
        }
        
        int fm = Math.max(first[n - 2], first[n - 3]);
        int sm = Math.max(second[n - 1], second[n - 2]);
        
        return Math.max(fm, sm);
    }
}